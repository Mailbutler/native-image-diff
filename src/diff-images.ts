import { __native_imageDiff } from "./native";

/**
 * Represents the dimensions of an image (width and height) in pixels.
 */
export interface Dimensions {
    /**
     * The width of this image in pixels.
     */
    width: number;
    /**
     * The height of this image in pixels.
     */
    height: number;
}
/**
 * A diffable image with a width and a height as well as a buffer containing the data.
 * The data must be in RGB or RGBA format.
 */
export interface DiffImage extends Dimensions {
    /**
     * A node buffer containing the image information in RGB or RGBA format.
     */
    data: Buffer;
}

/**
 * The result of diffing two images.
 */
export interface DiffResult {
    /**
     * The amount of pixels that didn't match.
     */
    pixels: number;
    /**
     * The output image visually representing the difference between two images.
     */
    image?: DiffImage;
    /**
     * A relative number which grows from `0` the higher the total difference between both images was.
     * This also takes the color differences between individual pixels into account.
     */
    totalDelta: number;
}

/**
 * Returns the amount of bytes per pixels for the given image.
 *
 * @param image The image of which the bytes per pixels should be calculated.
 *
 * @return The amount of bytes per pixels for the provided image.
 */
function getBytesPerPixel(image: DiffImage): number {
    const { data, width, height } = image;
    return data.length / (width * height);
}

/**
 * Checks that the given object is a valid `DiffImage`.
 *
 * @see DiffImage
 *
 * @return Whether the given image is of type `DiffImage`.
 */
function isDiffImage(image: DiffImage): image is DiffImage {
    if (typeof image !== "object") { return false; }
    const { data, width, height } = image;
    if (!Buffer.isBuffer(data)) { return false; }
    if (!Number.isInteger(width)) { return false; }
    if (!Number.isInteger(height)) { return false; }
    const bytesPerPixel = getBytesPerPixel(image);
    if (bytesPerPixel !== 3 && bytesPerPixel !== 4) { return false; }
    return true;
}

/**
 * Calculate the common dimensions of two images. This will return the maximum width and height of both provided images.
 *
 * @param image1 The first image of which to calculate the wdth and height.
 * @param image2 The second image of which to calculate the wdth and height.
 *
 * @return The common maximum dimensions of both images.
 */
function getCommonDimensions(image1: DiffImage, image2: DiffImage): Dimensions {
    return {
        width: Math.max(image1.width, image2.width),
        height: Math.max(image1.height, image2.height),
    };
}

export interface DiffImagesArguments {
    /**
     * The first image to compare.
     */
    image1: DiffImage;
    /**
     * The second image to compare.
     */
    image2: DiffImage;
    /**
     * A threshold of perceived color difference between two pixels. Ranging from `0` to `1`
     * with `0` being the most sensitive and `1` ignoring basically any difference between two pixels.
     * This threshold is applied to each pixel. Defaults to `0.1`.
     */
    colorThreshold?: number;
    /**
     * Whether to attempt to detect antialiasing when comparing the images. Defaults to `true`.
     */
    checkForAntialiasing?: boolean;
    /**
     * If set to `false` no image visualizing the difference is generated.
     */
    generateDiffImage?: boolean;
}

/**
 * Compare two images and return the result including meta information and a diff image as a buffer.
 *
 * @return The information about the difference between the two images.
 */
export function diffImages({
    image1,
    image2,
    colorThreshold,
    checkForAntialiasing,
    generateDiffImage,
}: DiffImagesArguments): DiffResult {
    // Make sure the input is valid.
    if (!isDiffImage(image1) || !isDiffImage(image2)) {
        throw new Error("Provided image wasn't a diffable image.");
    }
    const safeColorThreshold = typeof colorThreshold === "number" ? colorThreshold : 0.1;
    const safeCheckForAntialiasing = typeof checkForAntialiasing === "boolean" ? checkForAntialiasing : true;
    const safeGenerateDiffImage = typeof generateDiffImage === "boolean" ? generateDiffImage : true;
    // Call the native module.
    const { pixels, totalDelta, imageData: data } = __native_imageDiff(
        image1.data,
        image1.width,
        image1.height,
        image2.data,
        image2.width,
        image2.height,
        safeColorThreshold,
        safeCheckForAntialiasing,
        safeGenerateDiffImage,
    );
    if (safeGenerateDiffImage) {
        const image = {
            ...getCommonDimensions(image1, image2),
            data,
        };
        return { pixels, totalDelta, image };
    }
    return { pixels, totalDelta };
}
