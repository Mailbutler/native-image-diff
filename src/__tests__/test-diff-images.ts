import { diffImages } from "../diff-images";
import { readPngFileSync, writePngFileSync } from "node-libpng";

describe("diffImages", () => {
    describe("red and blue rectangles without alpha channel", () => {
        const image1 = readPngFileSync(`${__dirname}/fixtures/blue-red-32x16-1.png`);
        const image2 = readPngFileSync(`${__dirname}/fixtures/blue-red-32x16-2.png`);

        it("diffs the two images as expected", () => {
            const { image, pixels, totalDelta } = diffImages({
                image1,
                image2,
            });
            expect(image.data).toEqual(readPngFileSync(`${__dirname}/diffs/red-and-blue.png`).data);
            expect(pixels).toBe(86);
            expect(totalDelta).toBe(1291946.25);
        });

        it("doesn't generate an output image if not desired by the user", () => {
            const { image, pixels, totalDelta } = diffImages({
                image1,
                image2,
                generateDiffImage: false,
            });
            expect(image).toBeUndefined();
            expect(pixels).toBe(86);
            expect(totalDelta).toBe(1291946.25);
        });
    });

    describe("red and blue rectangles with alpha channel", () => {
        const image1 = readPngFileSync(`${__dirname}/fixtures/blue-red-green-alpha-32x16-1.png`);
        const image2 = readPngFileSync(`${__dirname}/fixtures/blue-red-green-alpha-32x16-2.png`);

        it("diffs the two images as expected", () => {
            const { image, pixels, totalDelta } = diffImages({
                image1,
                image2,
            });
            expect(image.data).toEqual(readPngFileSync(`${__dirname}/diffs/red-and-blue-with-alpha.png`).data);
            expect(pixels).toBe(81);
            expect(totalDelta).toBe(956607.1875);
        });
    });

    describe("with an antialiased image", () => {
        const image1 = readPngFileSync(`${__dirname}/fixtures/antialiasing-75x75-1.png`);
        const image2 = readPngFileSync(`${__dirname}/fixtures/antialiasing-75x75-2.png`);

        it("with antialiasing detection enabled", () => {
            const { image, pixels, totalDelta } = diffImages({
                image1,
                image2,
                checkForAntialiasing: true,
            });
            expect(image.data).toEqual(readPngFileSync(`${__dirname}/diffs/antialiasing-enabled.png`).data);
            expect(totalDelta).toBe(0);
            expect(pixels).toBe(0);
        });

        it("with antialiasing detection disabled", () => {
            const { image, pixels, totalDelta } = diffImages({
                image1,
                image2,
                checkForAntialiasing: false,
            });
            expect(image.data).toEqual(readPngFileSync(`${__dirname}/diffs/antialiasing-disabled.png`).data);
            expect(totalDelta).toBe(788076.6875);
            expect(pixels).toBe(265);
        });
    });

    describe("with different sizes", () => {
        const image1 = readPngFileSync(`${__dirname}/fixtures/different-sizes-20x20-1.png`);
        const image2 = readPngFileSync(`${__dirname}/fixtures/different-sizes-30x25-2.png`);

        it("detects the difference", () => {
            const { image, pixels, totalDelta } = diffImages({ image1, image2 });
            expect(image.data).toEqual(readPngFileSync(`${__dirname}/diffs/different-sizes.png`).data);
            expect(pixels).toBe(350);
            expect(totalDelta).toBe(12325250);
        });
    });

    describe("with different thresholds", () => {
        const image1 = readPngFileSync(`${__dirname}/fixtures/threshold-red-100-4x4-1.png`);
        const image2 = readPngFileSync(`${__dirname}/fixtures/threshold-red-110-4x4-2.png`);

        it("detects no differing pixels with a threshold of 0.05", () => {
            const { image, pixels, totalDelta } = diffImages({
                image1,
                image2,
                colorThreshold: 0.05,
            });
            expect(totalDelta).toBe(0);
            expect(pixels).toBe(0);
            for (let i = 0; i < image.data.length; ++i) { expect(image.data[i]).toBe(0); }
        });

        it("detects a difference with a threshold of 0.01", () => {
            const { image, pixels, totalDelta } = diffImages({
                image1,
                image2,
                colorThreshold: 0.01,
            });
            expect(image.data).toEqual(readPngFileSync(`${__dirname}/diffs/threshold-0.01.png`).data);
            expect(totalDelta).toBe(64.0384521484375);
            expect(pixels).toBe(4);
        });
    });

    describe("with invalid inputs", () => {
        it("throws an error if one of the inputs didn't have a buffer as data", () => {
            expect(() => {
                diffImages({
                    image1: {
                        data: "a string",
                        width: 10,
                        height: 10,
                    },
                    image2: {
                        data: Buffer.alloc(300),
                        width: 10,
                        height: 10,
                    },
                } as any);
            }).toThrowErrorMatchingSnapshot();
        });

        it("throws an error if one of the inputs didn't have a valid width", () => {
            expect(() => {
                diffImages({
                    image1: {
                        data: Buffer.alloc(300),
                        width: "ten",
                        height: 10,
                    },
                    image2: {
                        data: Buffer.alloc(300),
                        width: 10,
                        height: 10,
                    },
                } as any);
            }).toThrowErrorMatchingSnapshot();
        });

        it("throws an error if one of the inputs didn't have a valid height", () => {
            expect(() => {
                diffImages({
                    image1: {
                        data: Buffer.alloc(300),
                        width: 10,
                        height: "ten",
                    },
                    image2: {
                        data: Buffer.alloc(300),
                        width: 10,
                        height: 10,
                    },
                } as any);
            }).toThrowErrorMatchingSnapshot();
        });

        it("throws an error if one of the inputs wasn't an object", () => {
            expect(() => {
                diffImages({
                    image1: "some image I found somewhere",
                    image2: {
                        data: Buffer.alloc(300),
                        width: 10,
                        height: 10,
                    },
                } as any);
            }).toThrowErrorMatchingSnapshot();
        });

        it("throws an error if one of the input buffers had an unexpected amount of bytes per pixel", () => {
            expect(() => {
                diffImages({
                    image1: {
                        data: Buffer.alloc(500),
                        width: 10,
                        height: 10,
                    },
                    image2: {
                        data: Buffer.alloc(300),
                        width: 10,
                        height: 10,
                    },
                } as any);
            }).toThrowErrorMatchingSnapshot();
        });
    });
});
