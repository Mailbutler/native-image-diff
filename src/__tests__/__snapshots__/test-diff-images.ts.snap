// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffImages with invalid inputs throws an error if one of the input buffers had an unexpected amount of bytes per pixel 1`] = `"Provided image wasn't a diffable image."`;

exports[`diffImages with invalid inputs throws an error if one of the inputs didn't have a buffer as data 1`] = `"Provided image wasn't a diffable image."`;

exports[`diffImages with invalid inputs throws an error if one of the inputs didn't have a valid height 1`] = `"Provided image wasn't a diffable image."`;

exports[`diffImages with invalid inputs throws an error if one of the inputs didn't have a valid width 1`] = `"Provided image wasn't a diffable image."`;

exports[`diffImages with invalid inputs throws an error if one of the inputs wasn't an object 1`] = `"Provided image wasn't a diffable image."`;
